core/string/ustring.cpp

```diff
@@ -304,7 +304,7 @@ Error String::parse_url(String &r_scheme, String &r_host, int &r_port, String &r
 	return OK;
 }
 
-void String::copy_from(const StrRange<char> &p_cstr) {
+void String::parse_latin1(const StrRange<char> &p_cstr) {
```

Comment:
Arguably this function is parsing ASCII and has undefined behavior for extended characters (`latin1`). I hope somebody with more knowledge of encodings can weigh in on what this function actually does and should be called like.

Reply:
I would not call this one parse, that's just a simple memcpy, no interpretation of the underlying format. I fyou really want the parse prefix, then probably ASCII then, there is higher chance that people will find it than latin 1.

Reply:
There is definitely an interpretation going on. If nothing else, it's "utf-32 but only the first 255 values". Though that's not what it's used like. More often than not, it's used to construct strings from c strings, which from a cursory search is undefined behavior? Not exactly something you want to "just copy from".

Reply:
I did some more digging. Apparently C strings are encoded by whatever `-fexec-charset` (GCC / Clang) or `/execution-charset` (MSVC) is set to. MSVC defaults to 'system locale encodings' while GCC / clang default to UTF-8. In any case, whatever we're doing with these strings by passing them into _this_ function is wrong, and only doesn't cause problems because of low-bit overlaps of the encodings...

Reply:
> Arguably this function is parsing ASCII and has undefined behavior for extended characters (latin1).

What's undefined? Every defined Latin-1 (ISO-8859-1) character correspond one-to-one to the same character in the Unicode Basic Latin and Latin-1 Supplement blocks.

> /execution-charset (MSVC) is set to

It's already set to UTF-8: https://github.com/godotengine/godot/blob/b9437c393871c377b5ece131afaf831e78dfb6f8/platform/windows/detect.py#L443

Reply:
Yes, I found this statement now too. Turns out we've been parsing `UTF-8` encoded string literals incorrectly using an ascii parsing function this whole time. The main reason we haven't run into trouble is because nobody's used utf-8 characters in a string literal yet....

Reply:
I've settled for renaming it to `parse_ascii` and failing for values above 127. The only other correct choice would be to use the `parse_utf8` function for the `String` constructor instead, but that would come with unnecessary slow-downs, since we aren't using UTF-8 string literals (so far) anyway.

Reply:
> Turns out we've been parsing UTF-8 encoded string literals incorrectly using an ascii parsing function this whole time.

I think it was a deliberate choice to avoid unnecessary parsing, string literals in the code are either ASCII or `U".."` literals when it's necessary.

> The main reason we haven't run into trouble is because they overlap until 127, and nobody's used non-ascii characters in a string literal yet....

This happened few times.

Reply:
Also, it is documented to be Latin-1 in the GDExtension interface, so changing it to only handle ASCII (which is fine for the engine code) is a breaking change:

https://github.com/godotengine/godot/blob/b9437c393871c377b5ece131afaf831e78dfb6f8/core/extension/gdextension_interface.h#L1584-L1593

Reply:
Ok, fine with me — so we don't change any behavior and call the function `parse_latin1`? This is what I had originally. 
It would still be incorrect for utf-8 string literals, which should probably get a dedicated `ascii` function in a follow-up PR.

Reply:
I guess we can have both `ascii` (to use with literals) and `latin-1` (for GDExtensions).
Or keep `latin-1` for GDExtensions (for compatibility and mark it as deprecated) and only use `parse_utf8` for literals (probably better).

Reply:
If the script can specify its latin1, then yeah, parse_latin1 makes a lot of sense. But you are right, a dedicated ascii one for in engine seems better, to avoid some silent errors. Maybe even switch to parse_utf8 for those, as static string are generally pretty small, should not be too much cycles wasted (and we have a [PR](https://github.com/godotengine/godot/pull/99826) to speed that up (shameless plug))

Reply:
I think an `ascii` function would be good to have as it would be the fastest way to parse string literals, especially since almost all of them are unlikely to use non-ascii characters anyway.

I _could_ add an `ascii` function in this PR, but I think it's out of scope. I think it's better to call the existing function `latin` and call it a day, and try to address the implications of the mismatch in a follow-up PR.

Reply:
I'm not sure we really need the `ascii` one though. The only error right now is interpreting utf8 as latin1 for static strings. So we should use `parse_utf8` there, but I don't know what would be the use of an `ascii` one that would not be covered by `latin1` ( or maybe I'm just not seeing it)

Reply:
Its use-case would be that it would be faster than `utf-8`, but unlike `latin1` it would not be incorrect for string literals, instead logging an error on values > 127. So it could be safely used for most string literals.

Reply:
Oh ok, I see. Well I would be in favor of being always correct and always use parse_utf8. If we define it as utf-8 in the compilation toolchain, then stick to it and parse utf8. This will probably not even make a remarkable speed difference, and if it did, then we can go speed it up again. But correctness seems more important than performance here.

Reply:
Let's discuss this in follow-up PRs. I think we can agree that the current implementation parses `latin1`, which should be enough to finish up this PR.